toc/rfc/rfc-draft-v2-app-manifests.md

```diff
@@ -0,0 +1,613 @@
+      metadata:
+        annotations:
+          cassava: package metadata
+    lifecycle:
+      type: buildpack
+      buildpacks:
+        - name: ruby_buildpack
+        - name: java_buildpack
+        - url: git.example.com/my_buildpack
+      stack:
+        name: cflinuxfs4
+    environment_variables:
+      VAR1: value1
+      VAR2: value2
+    features:
+      ssh: true
+    metadata:
+      annotations:
+        contact: bob@example.com jane@example.com
+      labels:
+        sensitive: true
+    processes:
+      - type: web
+        command: start-web.sh
+        scale:
+          disk: 512M
+          instances: 3
+          memory: 500M
+        health_checks:
+          liveness:
+            http_endpoint: /health_check
+            type: http
+            invocation_timeout: 10
+            timeout: 10
+          readiness:
+            type: port
+        metadata:
+          annotations:
+            taro: process metadata
+      - type: worker
+        command: start-worker.sh
+        scale:
+          disk: 1G
+          instances: 2
+          memory: 256M
+        health_checks:
+          liveness:
+            type: process
+            timeout: 15
+  - name: app2
+    state: STOPPED
+    lifecycle:
+      type: docker
+    package:
+      docker:
+        image: docker-image-repository/docker-image-name
+        username: docker-user-name
+    processes:
+      - type: web
+        scale:
+          instances: 1
+          memory: 256M
+    sidecars:
+      - name: authenticator
+        processes:
+          - type: web
+          - type: worker
+        command: bundle exec run-authenticator
+        scale:
+          memory: 800M
+```
+
+#### Notable Changes
+
+**Routes**
+- Move `routes` to top-level
+- Route mappings are now "route-centric". The mappings are nested under routes,
+instead of under apps.
+- Add support for routing to non-`web` processes
+- Add support for routing to custom ports
+- Add metadata
+
+**Service Instances**
+- Move `service_instances` to top-level
+- Add support for provisioning/updating service instances (previously, only
+binding to existing service instances was supported)
+
+**Service Bindings**
+- Move `service_bindings` to top-level
+- Add support for route bindings
+- Add support for service keys
+- Add metadata
+
+**Apps**
+- Add `package`, `lifecycle`, `processes.scale`, and `processes.health_checks` to organize nodes
+- Add support for app features
+- Add support for app state
+- Add metadata to apps and nested app subresources
+
+### Future Extensions
+
+What follows are some possible extensions for v2 manifests. These demonstrate
+some of the future capabilities that could be enabled by v2 manifests. They are
+not currently possible with v1 manifests, so they would be new features only
+available with v2 manifests. These extensions are less-developed than the core
+v2 manifest design, and may change significantly prior to implementation.
+
+#### Merge State
```

Comment:
I don't have a strong opinion here but I also tend to a behaviour which is safe by default. It should be definitely possible to delete the resources with the manifest e.g. by defining a `merge` strategy in the manifest to `prune` but the default here could be `merge`. With this people could still have their app manifest and could have an extra manifest for the other space resource like `services` which usually are shared between the apps in a space.

Reply:
thinking further about this most probably there should be a different delete strategy for resources referenced by an application like a `service binding` or a `sidecar`. They should be deleted when they are not referenced any more by the application because they are "part" of the application definition. E.g. it is painful to delete a `sidecar` now days because you have to go over CF APIs. There is no support in the CF CLI or app manifest for this. Would be interested to see more comments on this.

Reply:
Yes, I'm also uncertain about how aggressive the current proposal is. 

My primary goal is to have something that can easily be explained. It's easier to explain "everything is declarative", vs "configuration of resources is declarative, but the set of resources themselves are not declarative, unless you pass a --prune flag", especially when the boundary between a "resource" and "configuration of a resource" is murky.

> Taking inspiration from Kubernetes: cf push on a full space manifest seems similar to kubectl apply of a multi-document YAML file or directory and it will only ever add & modify resources by default. Deletion is supported via [--prune](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#alternative-kubectl-apply-f-directory-prune) and has to be explicitly set.

Yea, the big distinction for me here is that k8s breaks things into multiple documents, so there is a clear boundary between what is declarative-by-default vs what is not. I could imagine a more revolutionary manifest proposal, where we do something similar for CF:

```yaml
---
kind: app
spec:
  name: my-app
---
kind: service_instance
spec:
  name: my-service-instance
```

The other inspiration is BOSH, which more-or-less has a single manifest that is strictly declarative.

> I would prefer if we go a similar route: resource themselves are fully declarative but a resource not listed should not be deleted implicitly following the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) (imagine the surprise of deleting a production database upon pushing a hello-world app to the same space).

Yes, this resonates with me. I'll think about this some more. I'm eager to hear additional feedback/suggested as well.

Reply:
> My primary goal is to have something that can easily be explained. It's easier to explain "everything is declarative", vs "configuration of resources is declarative, but the set of resources themselves are not declarative, unless you pass a --prune flag", especially when the boundary between a "resource" and "configuration of a resource" is murky.

I agree, this needs careful consideration and there would need to be a strong convention to prevent the varying behaviours we have in v1 manifests.

> > Taking inspiration from Kubernetes: cf push on a full space manifest seems similar to kubectl apply of a multi-document YAML file or directory and it will only ever add & modify resources by default. Deletion is supported via [--prune](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#alternative-kubectl-apply-f-directory-prune) and has to be explicitly set.
> 
> Yea, the big distinction for me here is that k8s breaks things into multiple documents, so there is a clear boundary between what is declarative-by-default vs what is not. I could imagine a more revolutionary manifest proposal, where we do something similar for CF: [...]
> The other inspiration is BOSH, which more-or-less has a single manifest that is strictly declarative.

The big difference to BOSH is that BOSH only has one manifest managing one resource. In that sense CF is probably closer to Kubernetes than BOSH. On the other hand moving towards Kubernetes-like primitives could also send the wrong signal to users.

Reply:
>Yea, the big distinction for me here is that k8s breaks things into multiple documents, so there is a clear boundary between what is declarative-by-default vs what is not. I could imagine a more revolutionary manifest proposal, where we do something similar for CF:

I was thinking in the same direction mentioning separate apps and services manifest. I think that this could simplify the understanding and reduce the perceived risk by CF users applying the new declarative behaviour because apps and services have different update lifecycle. Most of the time when you want to update your app most probably you don't want to update your service instance and having that in the same manifest or merged you relay on very good manifest diff and transparent merge behaviour which is not so trivial to implement.

Reply:
This is similar to how BOSH manages the deployments by manifest name. One thing to consider with this approach is how resource definitions could be moved between app manifests belonging to the same space  (this requirement will come sooner or later :-))

Reply:
The motivating usecase that we're envisioning here is an upstream continuous deployment system that assembles desired state upstream, builds a space-level manifest, and applies it to CF, ideally without requiring additional, imperative commands. We're probably over-fitting this design to that vision.

---

Another idea to consider: What if we encoded resource-by-resource "pruning" into the manifest itself. Something like:

```yaml
---
apps:
  meta:
    prune: true
  resources:
    - name: my-app
service_instances::
  meta:
    prune: false #default
  resources:
    - name: my-service-instance
```

Reply:
I think this is the best proposal until now because:
* people can organize the manifests belonging to a space as they are use to with v1
* resource definitions could be moved between manifests 
* default behaviour close to v1
* documentation should be straightforward

Reply:
The equivalent of how `kubectl` handles this would be something like:

```cf push -f manifest.yml --prune=apps,routes```

which could be backed by something like:

```POST /v3/spaces/:guid/actions/apply_manifest?prune=apps,routes```

---

I have mixed feelings about encoding the pruning behavior in the manifest itself.

Pros to putting it in the manifest:
1. The pruning behavior is "denormalized" to be close to the resource being pruned. That way it's clear by reading the manifest both what _will_ be pruned and what _could_ be pruned (e.g. what is a "resource" vs "configuration of a resource")
2. It's not necessary for a deployment system to split configuration between the manifest and meta-config (CLI flags, query params, etc).

Cons to putting it in the manifest:
1. It adds another layer of config (adding the `resources` layer in the default case).
2. The pruning behavior (probably) won't be included in the output of `cf create-space-manifest`, which breaks the symmetry between generating and applying manifests.
3. Developers may want to apply the same manifest, without modification, to different environments, with different pruning behavior (but maybe this could be solved with variable substitution?).

Reply:
Tim also pointed me to [this blog post](https://www.meshcloud.io/en/blog/how-to-implement-declarative-deletion/), which is an interesting industry survey of declarative deletion.

Reply:
Thanks for the blog post link! A really good summary for declarative deletion. Based on the experience shared in the blog it reads to me that a well supported imperative deletion should be our priority.  Extending that to the app manifest could be done in a later point if desired by users. Regarding the imperative approach if we want to have a resource-by-resource "pruning" my understanding is that we will need something like:

`cf push -f manifest.yml --prune_apps=app-1,app-2 --prune_routes=route-1,route-2`

and on API side:

`POST /v3/spaces/:guid/actions/apply_manifest?prune_apps=app-1,app-2&prune_routes=route-1,route-2`

@Gerg Is that what you meant in your example above?

Reply:
My example was inspired by on how `kubectl apply`'s `--prune-allowlist` flag works [[docs](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#alternative-kubectl-apply-f-directory-prune)]. You give it a list of the kinds of resource you want to delete. For example:

```kubectl apply --prune -f manifest.yaml --all --prune-allowlist=core/v1/ConfigMap```

would prune all excess `ConfigMaps`, but leave other kinds of resources alone.

My intention was for:
```
cf push -f manifest.yml --prune=apps,routes
```

to prune excess apps and routes, but leave other resources alone.

---

Regarding imperative deletion, I think we already have that well covered via the `cf delete-*` commands. For example, if the developer already has a list of specific routes to delete, it doesn't seem much harder to pass those apps to `cf delete-route` vs a `--prune-routes` flag.

Reply:
I was looking into this from the perspective where one space is managed by multiple manifests and then the approach with `cf push -f manifest.yml --prune=apps,routes` doesn't fit so well but I can see that for the use case you described above:

>The motivating usecase that we're designing around is an upstream continuous deployment system that assembles desired state upstream, builds a space-level manifest, and applies it to CF, ideally without requiring imperative commands. We're probably over-fitting this design to that vision. 

it makes sense. 

I agree with:
>Regarding imperative deletion, I think we already have that well covered via the cf delete-* commands.

Reply:
I substantially updated the declarative resource deletion section in 17db5d1, based on this thread. Let me know what you think.